src/clj_gatling/simulation.clj

```diff
@@ -123,62 +137,90 @@
-                                                                                   @sent-requests
-                                                                                   simulation-start
-                                                                                   (LocalDateTime/now))
-                                   target-concurrency (* concurrency (modifier-fn {:progress progress
-                                                                                   :duration duration
-                                                                                   :context context}))]
-                               (> target-concurrency @concurrent-scenarios)))
-                          (constantly true))]
-    (go-loop []
-      (if (should-run-now?)
-        (do
-          (swap! concurrent-scenarios inc)
-          (let [result (<! (run-scenario-once options scenario user-id))]
+  (let [c                (async/chan)
+        continue?        (continue-fn {:runner runner
+                                       :simulation-start simulation-start
+                                       :force-stop force-stop})
+        should-run-now?  (if concurrency-distribution
+                           (let [modifier-fn (if (= 2 (arg-count concurrency-distribution))
+                                               (fn [{:keys [progress context]}]
+                                                 (concurrency-distribution progress context))
+                                               concurrency-distribution)]
+                             #(let [[progress duration] (runners/calculate-progress runner
+                                                                                    (:sent @simulation-requests)
+                                                                                    @simulation-start
+                                                                                    (LocalDateTime/now))
+                                    target-concurrency (* concurrency (modifier-fn {:progress progress
+                                                                                    :duration duration
+                                                                                    :context context}))]
+                                (atomic-concurrency-check target-concurrency concurrent-scenarios)))
+                           #(atomic-concurrency-check concurrency concurrent-scenarios))
+        starting-context (scenario-context options scenario user-id)]
+    (swap! init-sync dec)
+    (go-loop [context starting-context]
+      (while (> @init-sync 0)
+        (<! (timers/timeout 100)))
+      (let [[old-total expected] (prepare-run scenario simulation-requests scenario-requests scenario-runs)]
+        (while (and (continue? old-total (LocalDateTime/now)) (not (should-run-now?)))
+          (<! (timers/timeout (rand-int 20))))
```

Comment:
Because we might get to this point and not actually want to run yet (enough runners already in progess), we spin for a short random time, then check again. This means we don't keep generating and throwing away `context`, which is what I had initially (just `recur` if we shouldn't run this time). This keeps the clj-gatling host load lower even with a heavyweight context generator and a lot of concurrency.